WalletWasabi/Services/WalletService.cs

```diff
@@ -966,300 +901,230 @@ public Operation(Script script, Money amount, string label)
+				builder.Send(request.Destination, amountRequest.Amount);
+				if (amountRequest.SubtractFee)
+				{
+					builder.SubtractFees();
+				}
+			}
 
-			Logger.LogInfo<WalletService>($"Fee: {fee.Satoshi} Satoshi.");
+			HdPubKey changeHdPubKey = null;
 
-			// 5. How much to spend?
-			long toSendAmountSumInSatoshis = toSend.Select(x => x.Amount).Sum(); // Does it work if I simply go with Money class here? Is that copied by reference of value?
-			var realToSend = new (Script script, Money amount, string label)[toSend.Length];
-			for (int i = 0; i < toSend.Length; i++) // clone
-			{
-				realToSend[i] = (
-					new Script(toSend[i].Script.ToString()),
-					new Money(toSend[i].Amount.Satoshi),
-					toSend[i].Label);
-			}
-			for (int i = 0; i < realToSend.Length; i++)
+			if (payments.TryGetCustomRequest(out DestinationRequest custChange))
 			{
-				if (realToSend[i].amount == Money.Zero) // means spend all
-				{
-					realToSend[i].amount = allowedSmartCoinInputs.Select(x => x.Amount).Sum();
-
-					realToSend[i].amount -= new Money(toSendAmountSumInSatoshis);
+				var changeScript = custChange.Destination.ScriptPubKey;
+				changeHdPubKey = KeyManager.GetKeyForScriptPubKey(changeScript);
 
-					if (subtractFeeFromAmountIndex is null)
-					{
-						realToSend[i].amount -= fee;
-					}
+				var changeStrategy = payments.ChangeStrategy;
+				if (changeStrategy == ChangeStrategy.Custom)
+				{
+					builder.SetChange(changeScript);
 				}
-
-				if (subtractFeeFromAmountIndex == i)
+				else if (changeStrategy == ChangeStrategy.AllRemainingCustom)
 				{
-					realToSend[i].amount -= fee;
+					builder.SendAllRemaining(changeScript);
 				}
-
-				if (realToSend[i].amount < Money.Zero)
+				else
 				{
-					throw new InsufficientBalanceException(fee + Money.Satoshis(1), realToSend[i].amount + fee);
+					throw new NotSupportedException(payments.ChangeStrategy.ToString());
 				}
 			}
+			else
+			{
+				KeyManager.AssertCleanKeysIndexed(isInternal: true);
+				KeyManager.AssertLockedInternalKeysIndexed(14);
+				changeHdPubKey = KeyManager.GetKeys(KeyState.Clean, true).RandomElement();
 
-			var toRemoveList = new List<(Script script, Money money, string label)>(realToSend);
-			toRemoveList.RemoveAll(x => x.money == Money.Zero);
-			realToSend = toRemoveList.ToArray();
+				builder.SetChange(changeHdPubKey.P2wpkhScript);
+			}
 
-			// 1. Get the possible changes.
-			Script changeScriptPubKey;
-			var sb = new StringBuilder();
-			foreach (var item in realToSend)
+			builder.SendEstimatedFees(feeRate);
+
+			var psbt = builder.BuildPSBT(false);
+
+			var spentCoins = psbt.Inputs.Select(txin => smartCoinsByOutpoint[txin.PrevOut]).ToArray();
+
+			var realToSend = payments.Requests.Select(t => (label: t.Label,
+												destination: t.Destination,
+												amount: psbt.Outputs.FirstOrDefault(o => o.ScriptPubKey == t.Destination.ScriptPubKey)?.Value))
+									.Where(i => i.amount != null)
+									.ToArray();
```

Comment:
The indentation can be improved in the above lines.

Reply:
Improved.